### ISSUE TITLE
Schrödinger's Train/Test Split

### ISSUE THREAD

**Issue by AlexanderNenninger** [pola-rs/polars#22862]
### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

You may need to run the code a a few times to see the error.

```python
import polars as pl

pl.show_versions()


def stratified_train_test_split(
    df: pl.DataFrame | pl.LazyFrame,
    group_by: str,
    label_column: str,
    aggregation: pl.Expr | None = None,
    test_ratio: float = 0.2,
    seed: int | None = None,
) -> dict[str, pl.DataFrame | pl.LazyFrame]:
    """Split a DataFrame into train and test sets with stratification."""

    df_lazy = df.lazy()

    maintain_order = seed is not None

    if aggregation is None:
        # Use the majority class for stratification
        aggregation = pl.col(label_column).mode().first()

    # Identify the majority class for each group
    group_classes = (
        df_lazy.select((group_by, label_column))
        .group_by(group_by, maintain_order=maintain_order)
        .agg(aggregation.alias("_majority_class"))
    )

    # Stratified split for each class
    stratified_groups = []

    # Get unique classes
    unique_classes = (
        group_classes.select("_majority_class")
        .unique(maintain_order=maintain_order)
        .collect()
        .to_series()
    )

    for class_value in unique_classes:
        # Get groups for this class
        class_groups = group_classes.filter(pl.col("_majority_class") == class_value)

        # Create deterministic seed per class
        class_seed = None if seed is None else seed + hash(str(class_value)) % 10000

        # Randomly assign test groups for this class
        assigned_groups = (
            class_groups.with_columns(
                pl.int_range(0, pl.len())
                .shuffle(seed=class_seed)
                .alias("_random_index")
            )
            .with_columns(
                (pl.col("_random_index") < (pl.len() * test_ratio)).alias("_is_test")
            )
            .drop("_random_index", "_majority_class")
        )

        stratified_groups.append(assigned_groups)

    # Combine all results
    all_groups = pl.concat(stratified_groups)

    # Join with original data
    result = df_lazy.join(
        all_groups,
        on=group_by,
        how="left",
        maintain_order="left" if seed is not None else None,
    )

    # Split into train and test
    train = result.filter(~pl.col("_is_test")).drop("_is_test")
    test = result.filter(pl.col("_is_test")).drop("_is_test")

    train = train.collect()
    test = test.collect()
    result = result.collect()

    return {"train": train, "test": test, "result": result}


groups = ["A", "A", "A", "A", "B", "B", "B", "B", "C", "C", "C", "C", "D", "D"]
classes = [0, 2, 2, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 2]
df = pl.DataFrame(
    {
        "group": groups,
        "class": classes,
        "value": range(len(groups)),
    }
)

assert len(df) == 14

result = stratified_train_test_split(
    df,
    group_by="group",
    label_column="class",
    test_ratio=0.33,
    seed=42,
    aggregation=pl.col("class").filter(pl.col("class") != 0).mode().first(),
)

assert result["train"].height + result["test"].height == result["result"].height, (
    "Wrong split with seed"
)

result = stratified_train_test_split(
    df,
    group_by="group",
    label_column="class",
    test_ratio=0.33,
    seed=None,
    aggregation=pl.col("class").filter(pl.col("class") != 0).mode().first(),
)

assert result["train"].height + result["test"].height == result["result"].height, (
    "Wrong split without seed"
)


```

### Log output

```shell
keys/aggregates are not partitionable: running default HASH AGGREGATION
found multiple sources; run comm_subplan_elim
cache nodes will be removed because predicates don't match
join parallel: true
UNION: union is run in parallel
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
LEFT join dataframes finished
dataframe filtered
found multiple sources; run comm_subplan_elim
cache nodes will be removed because predicates don't match
join parallel: true
UNION: union is run in parallel
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
LEFT join dataframes finished
dataframe filtered
found multiple sources; run comm_subplan_elim
cache nodes will be removed because predicates don't match
join parallel: true
UNION: union is run in parallel
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
LEFT join dataframes finished
keys/aggregates are not partitionable: running default HASH AGGREGATION
found multiple sources; run comm_subplan_elim
cache nodes will be removed because predicates don't match
join parallel: true
UNION: union is run in parallel
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
LEFT join dataframes finished
dataframe filtered
found multiple sources; run comm_subplan_elim
cache nodes will be removed because predicates don't match
join parallel: true
UNION: union is run in parallel
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
LEFT join dataframes finished
dataframe filtered
found multiple sources; run comm_subplan_elim
cache nodes will be removed because predicates don't match
join parallel: true
UNION: union is run in parallel
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
dataframe filtered
LEFT join dataframes finished
```

### Issue description

Schrödinger's Train/Test Splits. Sometimes a group ends up in both, train and test dataset, even though, they have a well-defined assignment in the  `result` dataframe. 

### Expected behavior

Groups are only assigned to one of train or test. 

### Installed versions

<details>

```
--------Version info---------
Polars:              1.30.0
Index type:          UInt32
Platform:            Linux-5.10.236-228.935.amzn2.x86_64-x86_64-with-glibc2.39
Python:              3.12.9 (main, Mar 10 2025, 16:51:21) [GCC 13.3.0]
LTS CPU:             False

----Optional dependencies----
Azure CLI            <not installed>
adbc_driver_manager  <not installed>
altair               5.5.0
azure.identity       <not installed>
boto3                1.37.3
cloudpickle          2.2.1
connectorx           <not installed>
deltalake            <not installed>
fastexcel            0.12.1
fsspec               2023.12.2
gevent               <not installed>
google.auth          2.40.1
great_tables         <not installed>
matplotlib           3.10.3
numpy                1.26.4
openpyxl             <not installed>
pandas               2.2.3
polars_cloud         <not installed>
pyarrow              19.0.1
pydantic             1.10.22
pyiceberg            <not installed>
sqlalchemy           2.0.41
torch                2.7.0+cu126
xlsx2csv             <not installed>
xlsxwriter           3.2.3
```

</details>

**Comment by cmdlineluser** [comment]
I think at the very least you would need 

```python
train, test, result = pl.collect_all([train, test, result])
```
 
Because anything with non-stable output (group_by, shuffle, join) will cause different results for each `collect`.

- https://docs.pola.rs/user-guide/lazy/multiplexing/

**Comment by MarcoGorelli** [comment]
closing as per https://github.com/pola-rs/polars/issues/22862#issuecomment-2898591934, but thanks for the report
